Index: src/groovy/grails/plugin/fsm/FsmSupport.groovy
===================================================================
--- src/groovy/grails/plugin/fsm/FsmSupport.groovy	(revision 75259)
+++ src/groovy/grails/plugin/fsm/FsmSupport.groovy	(working copy)
@@ -125,9 +125,10 @@
     {
         def fromState = a_grammar.fromState
         def toState = a_grammar.toState
-        if (!actions[fromState])
-        	actions[fromState] = [:]
-        actions[fromState][toState] = a_grammar.action
+        def key = "${a_grammar.event}-${fromState}"
+        if (!actions[key])
+        	actions[key] = [:]
+        actions[key][toState] = a_grammar.action
     }
 
     def reset()
@@ -137,7 +138,7 @@
 
     def fire(a_event)
     {
-        assert currentState, "Invalid current state '${currentState}': pass into constructor"
+       assert currentState, "Invalid current state '${currentState}': pass into constructor"
         assert transitions.containsKey(a_event), "Invalid event '${a_event}', should be one of ${transitions.keySet()}"
 
         def fromState = currentState
@@ -164,8 +165,9 @@
         }
 
         // Trigger the action defined if any
-        if (actions[fromState] && actions[fromState][currentState]) {
-        	def act = actions[fromState][currentState]
+        def key = "${a_event}-${fromState}"
+        if (actions[key] && actions[key][currentState]) {
+        	def act = actions[key][currentState]
             act.setDelegate(target)
             act()
         }
Index: FsmGrailsPlugin.groovy
===================================================================
--- FsmGrailsPlugin.groovy	(revision 75259)
+++ FsmGrailsPlugin.groovy	(working copy)
@@ -17,7 +17,7 @@
 
 class FsmGrailsPlugin {
     // the plugin version
-    def version = "0.6.4"
+    def version = "0.6.5"
     // the version or versions of Grails the plugin is designed for
     def grailsVersion = "1.1.1 > *"
     // the other plugins this plugin depends on
Index: grails-app/domain/FsmSupportDummy.groovy
===================================================================
--- grails-app/domain/FsmSupportDummy.groovy	(revision 75259)
+++ grails-app/domain/FsmSupportDummy.groovy	(working copy)
@@ -7,6 +7,7 @@
     def name
     def status = 'initiala'
     String mood = 'ToBeOverridenInFsmDef'
+    def cheers
 
     def amount  // A number > 0 will allow the status flow to start
 
@@ -62,6 +63,16 @@
                     from('stopped').to('finished')
                     from('running').to('finished')
                 }
+                flow.on('hello') {
+                    from('stopped').to('stopped').act({
+                        cheers = 'helloooou'
+                    })
+                }
+                flow.on('goodbye') {
+                    from('stopped').to('stopped').act({
+                        cheers = 'goodbyeee'
+                    })
+                }
             }
         ]
     ]
@@ -76,4 +87,5 @@
     def changeName(n) {
     	name = n
     }
+
 }
Index: test/unit/FsmSupportUnitTests.groovy
===================================================================
--- test/unit/FsmSupportUnitTests.groovy	(revision 75259)
+++ test/unit/FsmSupportUnitTests.groovy	(working copy)
@@ -83,5 +83,14 @@
         println("MOOD: ${foo.mood}")
         assert foo.mood == 'low'
     }
+
+    void testDifferentEventsThrowDifferentWorkflows() {
+        FsmSupportDummy foo = new FsmSupportDummy()
+        foo.status = 'stopped'
+        foo.fire('status', 'hello')
+        assert foo.cheers == 'helloooou'
+        foo.fire('status', 'goodbye')
+        assert foo.cheers == 'goodbyeee'
+    }
 }
 
